### PR TITLE
Refactor preparing for JSON-LD

### PR DESCRIPTION
**Pull request by woutslakhorst**
JSON-LD requires context to parse documents. Therefore logic has to be removed from Document and FieldIndexer and added to Collection and Index.

Also:
- added Scalar for better value comparing and less error checking
- fixed a nasty infinite loop bug

### CODE REVIEW

File: index_part.go [hunk 1]
```diff
@@ -72,63 +86,27 @@ func (j fieldIndexer) Name() string {
...
 	}
 
-	if s, ok := value.(string); ok {
+	if s, ok := scalar.value.(string); ok {
 		tokens := j.tokenizer(s)
-		result := make([]interface{}, len(tokens))
+		result := make([]Scalar, len(tokens))
 		for i, t := range tokens {
-			result[i] = t
+			result[i] = ScalarMustParse(t)
```
Comment (reinkrul) [inline index_part.go:102]: wouldn't it make more sense to make `Scalar` an interface and have implementing structs for each supported type?

File: search.go [hunk 2]
```diff
@@ -41,11 +41,11 @@ type QueryPart interface {
 	Name() string
 
 	// Seek returns the key for cursor.Seek
-	Seek() (Key, error)
+	Seek() Scalar
```
Comment (reinkrul) [inline search.go:44]: why does it return a scalar (or anything at all)? Seek is usually an operation (at least when thinking of spinning/sequenced storage) to move a cursor to a set position?

File: types.go [hunk 3]
```diff
@@ -42,21 +48,52 @@ func (r Reference) ByteSize() int {
...
+// Scalar represents a JSON or JSON-LD scalar (string, number, true or false)
+type Scalar struct {
+	value interface{}
+}
+
+// ErrInvalidValue is returned when an invalid value is parsed
+var ErrInvalidValue = errors.New("invalid value")
+
+// ScalarParse returns a Scalar based on an interface value. It returns ErrInvalidValue for unsupported values.
+func ScalarParse(value interface{}) (Scalar, error) {
```
Comment (reinkrul) [inline types.go:60]: `Verb-Type` is usually a nicer format (thus `ParseScalar`)?

File: types.go [hunk 4]
```diff
@@ -42,21 +48,52 @@ func (r Reference) ByteSize() int {
...
-		var buf [4]byte
-		binary.BigEndian.PutUint32(buf[:], castData)
-		return buf[:], nil
+// Scalar represents a JSON or JSON-LD scalar (string, number, true or false)
+type Scalar struct {
+	value interface{}
+}
+
+// ErrInvalidValue is returned when an invalid value is parsed
+var ErrInvalidValue = errors.New("invalid value")
```
Comment (reinkrul) [inline types.go:57]: Unsupported value.

Maybe add provided type? (like `fmt.Errorf("unsupported value: %T", value)`)

File: types.go [hunk 5]
```diff
@@ -42,21 +48,52 @@ func (r Reference) ByteSize() int {
...
+		if castData {
+			return []byte{1}
+		}
+		return []byte{0}
 	case string:
-		return []byte(castData), nil
+		return []byte(castData)
 	case float64:
 		var buf [8]byte
 		binary.BigEndian.PutUint64(buf[:], math.Float64bits(castData))
```
Comment (reinkrul) [inline types.go:94]: didn't see this before, does it work for negative numbers?

File: search.go (hunk 2 above)
Comment (woutslakhorst) [reply search.go:44]: it returns the seek value where the cursor needs to move to.

File: types.go (hunk 4 above)
Comment (woutslakhorst) [reply types.go:57]: that would be for the caller

File: types.go (hunk 5 above)
Comment (woutslakhorst) [reply types.go:94]: added test for it

File: index_part.go (hunk 1 above)
Comment (woutslakhorst) [reply index_part.go:102]: changed it, leads to more code but less switch statements

**Review by reinkrul — APPROVED** [review]

**Pull request merged**